Decorate exceptions with backtrace information (#330)
ghc-proposals/ghc-proposals: proposals/0000-exception-backtraces.rst

* DWARF debug information in conjunction with GHC's `built-in stack unwinder
  <https://www.haskell.org/ghc/blog/20200405-dwarf-3.html>`_ can be used
  to provide terse (but nevertheless useful) backtraces with no runtime
  overhead in the non-failing case (although backtrace collection is quite
  slow)

All of these backtrace options have their time and place, offering a range of
levels of detail, executable size, and runtime overhead. GHC, being a compiler,
shouldn't be in the business of dictating which of these mechanisms should be
used to report exception backtraces.  Consequently, our ``Backtrace`` type is
designed to capture them all: ::

    -- | An exception backtrace.
    data Backtrace
        = CostCenterBacktrace (Ptr GHC.Stack.CCS.CostCentreStack)
          -- ^ a cost center profiler backtrace
        | HasCallStackBacktrace GHC.Stack.CallStack
          -- ^ a stack from 'GHC.Stack.HasCallStack'
        | ExecutionBacktrace [GHC.ExecutionStack.Location]
          -- ^ a stack unwinding (e.g. DWARF) backtrace

    instance Show Backtrace

With such a type we can easily write a variant of ``throwIO`` that, for

> instance, attaches a DWARF backtrace: ::

Reviewer: ```suggestion
instance, attaches a ``HasCallStack`` backtrace: ::
```

>     -- | Throws an exception with a 'HasCallStack' backtrace.

Reviewer 2: Isn't this a `HasCallStackBacktrace`, not a `ExecutionBacktrace` as implied above?  Also it's not clear how the Backtrace type enables this, at least not without a definition for `throw`.

global backtrace mechanism from the environment: ::

    setBacktraceMechanismFromEnv :: IO ()
    setBacktraceMechanismFromEnv =
        getEnv "GHC_BACKTRACE" >>= setGlobalBacktraceMechanism . parseBacktraceMechanism

This could then be called during program initialization, providing the ease of
configuration found in other languages.

Effect and Interactions
-----------------------

This mechanism provides users with a convenient means of gaining greater
insight into the sources of exceptions. Currently the runtime system's ``+RTS
-xc`` flag provides an ad-hoc mechanism in the runtime system which relies on the
cost-center profiler. In principle the ``-xc`` mechanism is subsumed by the
mechanism proposed here.

Costs and Drawbacks
-------------------

While the global backtrace mechanism is convenient, it suffers from the usual
drawbacks associated with global state: it does not compose well and may result
in surprising behavior when manipulated by more than one actor.

This being said, we consider this approach to be a compromise which reflects
the fact that stack traces are primarily a debugging tool and somewhat of a

> cross-cutting concern. While a stateful approach would be preferred, we believe

Reviewer 3: ```suggestion
cross-cutting concern. While a stateless approach would be preferred, we believe
```

Author: Thanks!

that this compromise is a significant improvement over the status quo.

Alternatives
------------

In addition to the runtime-configurable ``setGlobalBacktraceMechanism``
mechanism described above, GHC could gain support for setting the backtrace
mechanism at compile-time via a compiler flag (this would essentially come down
to GHC emitting a call to ``setGlobalBacktraceMechanism`` in its start-up
code).

The ``Backtrace`` type described above precludes use of more than one backtrace
mechanism at once. It's possible that we might want a more general definition
which allows use of more than one mechanism at once.

Alternatively, the community might rather choose one of the backtrace

> mechanisms discussed about and use this mechanism exclusively in exception

Reviewer: ```suggestion
mechanisms discussed above and use this mechanism exclusively in exception
```

Making backtraces ubiquitous
~~~~~~~~~~~~~~~~~~~~~~~~~~~~

While these functions are useful building blocks, they do not
address the most common debugging scenario: searching for an exception
thrown by a third-party library (which likely calls the old ``throw`` and
consequently would not produce a backtrace). For this we propose a pragmatic,
stateful approach to allow the user to select which mechanism should be used
for backtrace collection in ``mkSomeExceptionWithLocation``: ::

    module GHC.Exception.Backtrace where

    -- | Which kind of backtrace to collect when an exception is thrown.
    data BacktraceMechanism
      = NoBacktrace
      | CollectCostCenterBacktrace
      | HasCallStackBacktrace
      | ExecutionStackBacktrace

    -- | Set the global backtrace mechanism.
    -- (this state would be represented internally as a simple global IORef.
    setGlobalBacktraceMechanism :: BacktraceMechanism -> IO ()

The ``collectBacktrace`` primitive referenced in the previous section then
simply dispatches to the appropriate backtrace collection scheme as determined
by the currently selected ``BacktraceMechanism``.

Examples
--------

> User programs would typically call ``setGlobalBacktraceMechanism`` during

Reviewer 4: I'm really nitpicking here. But I don't think “global backtrace mechanism” characterises this option so well. Rather, I would think of it more of a “default backtrack mechanism”.

Author: That is a fair point.

Reviewer 5: @aspiwack , @bgamari ,

Is this agreed?

Should I use `setDefaultBacktraceMechanisms` as name for this function? (I've added a plural "s" to the name, because one can set a list of mechanisms.)

Reviewer 4: It's what the proposal currently says. So it's probably agreed :slightly_smiling_face:

in surprising behavior when manipulated by more than one actor.

This being said, we consider this approach to be a compromise which reflects
the fact that stack traces are primarily a debugging tool and somewhat of a
cross-cutting concern. While a stateless approach would be preferred, we
believe that this compromise is a significant improvement over the status quo.

Alternatives
------------

In addition to the runtime-configurable ``setGlobalBacktraceMechanism``
mechanism described above, GHC could gain support for setting the backtrace
mechanism at compile-time via a compiler flag (this would essentially come down
to GHC emitting a call to ``setGlobalBacktraceMechanism`` in its start-up
code).

The ``Backtrace`` type described above precludes use of more than one backtrace
mechanism at once. It's possible that we might want a more general definition
which allows use of more than one mechanism at once.

Alternatively, the community might rather choose one of the backtrace
mechanisms discussed above and use this mechanism exclusively in exception
backtraces. However, we suspect that a single mechanism won't be sufficient:

> * there have been `previous efforts
>    <https://gitlab.haskell.org/ghc/ghc/issues/17040>`_ to add ``HasCallStack``

Reviewer 4: ```suggestion
* there have been `previous efforts <https://gitlab.haskell.org/ghc/ghc/issues/17040>`_ to add ``HasCallStack``
```
(the additional newline messes up with the rendering in Github)

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

All exceptions in Haskell are currently represented by the
``Control.Exception.SomeException`` type which captures the exception along
with its ``Exception`` dictionary: ::

    data SomeException where
        SomeException :: forall a. (Exception a)
                      => a -> SomeException

Such a ``SomeException`` value carries all of the information needed to catch,
identify, and display an exception, but lacks any generic facility for encoding
relevant metadata about the exception.

As a first step towards attaching backtraces to exceptions, we propose to extend
``SomeException`` to carry an additional "context" value, consisting of a collection
of dynamically-typed annotations. The dynamic nature of these annotations
allows the mechanism to handle not just backtraces but any arbitrary
context-relevant value which may be inspected by a handler for analysis and logging.
To avoid breaking existing users of ``SomeException``, we introduce this
context as an implicit parameter constraint:  ::

    data SomeException where
        SomeException :: forall a. (Exception a, ?context :: ExceptionContext)
                      => a -> SomeException

>     data ExceptionContext = ExceptionContext [SomeExceptionAnnotation]

Author: A call stack is a single annotation. Until now I had been considered `ExceptionContext` to be an unordered bag of annotations and indeed through that lens the ordered natured of list is a bit odd. One way of addressing this would be to state that `ExceptionContext`s are to be interpreted as *stacks* of annotations (which would roughly reflect the lexical scoping of the handlers which added the annotations). This isn't *strictly* true since multiple annotations may be added by the same call to, e.g., `withExceptionContext`. However, I think it's true enough to be a useful convention and at least somewhat resolve the semantic issue that you raise.

 * 

 I do agree that semantically the ordered nature of the list representation is somewhat odd as `ExceptionContext` is really an unordered bag of annotations. Note that it's not necessarily a map; there may be more than one `HasCallStackBacktrace` in the However, practically-speaking list is one of the few structures which we can readily use in `base`; I am not sure that "somewhat odd" really justifies adding a map implementation

``Exception`` class be extended with a new ``toExceptionWithContext`` method: ::

    class (Typeable e, Show e) => Exception e where
        -- These are unchanged:
        toException            :: e -> SomeException
        fromException          :: SomeException -> Maybe e
        displayException       :: e -> String

        -- This is new:
        toExceptionWithContext :: e -> ExceptionContext -> SomeException
        toExceptionWithContext e ?exceptionContext = SomeException e

        -- toException is implemented in terms of toExceptionWithContext
        toException e = toExceptionWithContext e mempty

The ``Exception`` instance for ``SomeException`` would accumulate and display
contexts: ::

    instance Exception SomeException where
        -- This is unchanged:
        toException se = se
        fromException = Just

        -- toExceptionWithContext *adds* context to an existing SomeException:
        toExceptionWithContext se@(SomeException e) ctxt =
            SomeException e
          where ?exc_context = ctxt <> exceptionContext se

        -- displayException shows context after the exception itself:
        displayException (SomeException e) =

>             displayException e ++ "\n" ++ displayExceptionContext ?exc_context

Author: Yes, this is true. However, I can see few good ways of addressing this. One way would be to add yet another method to `Exception`:
```
class Exception a where
    ...
    displayException :: a -> String
    displayException = show

    displayExceptionWithContext :: a -> String
    displayExceptionWithContext x = displayException x ++ "\n" ++ displayExceptionContext ?exc_context
```
`displayExceptionWithContext` would then become the preferred function for formatting exceptions for the user. However, this merely moves the problem: instead of needing to update all *definition* sites of `displayException` you now rather need to update all call-sites. A rather [unscientific poll of Hackage](https://hackage-search.serokell.io/?q=displayException) suggests that call-sites may actually be more numerous than method definitions, so I suspect that this change would actually be worse than the current proposal.

>             displayException e ++ "\n" ++ displayExceptionContext ?exc_context

Reviewer 6: Where is `?exc_context` brought into scope here? I think, without a `SomeException` wrapper, we just won't have context. Which isn't necessarily a problem, *generally* if you have a `SomeException` it's because you don't know more about what you want to be catching.

>             displayException e ++ "\n" ++ displayExceptionContext ?exc_context

Author: Yes, quite right. I don't know what I was thinking when I wrote this.

    displayExceptionContext :: ExceptionContext -> String
    displayExceptionContext = ...

Since the ``SomeException``'s ``displayException`` implementation is used to
by GHC's top-level exception handler to display uncaught exceptions, this
change carries the consequence that uncaught exceptions will have their context
automatically printed as part of the error message presented to the user.

To make context-carrying exceptions easier to work with, we propose to
introduce the following combinators: ::

    -- In Control.Exception:
    exceptionContext :: SomeException -> ExceptionContext
    exceptionContext (SomeException _) = ?exceptionContext

    addContext :: ExceptionContext -> SomeException -> SomeException

    catchWithContext   :: Exception e => m a -> (e -> ExceptionContext -> m a) -> m a

>     handleWithContext  :: Exception e => (e -> ExceptionContext -> m a) -> m a -> m a

Author: Yes, I quite like this.

>     handleWithContext  :: Exception e => (e -> ExceptionContext -> m a) -> m a -> m a

Author: Indeed, this is a nice addition which I have folded into the proposal.

which mechanism(s) should be used for backtrace collection in ``throw``,
``throwIO`` and similar functions: ::

    module GHC.Exception.Backtrace where

    -- | Which kind of backtrace to collect when an exception is thrown.
    data BacktraceMechanism
      = -- | collect a cost center stacktrace (only available when built with profiling)
        CostCenterBacktraceMech
      | -- | use execution stack unwinding with given limit
        ExecutionStackBacktraceMech
      | -- | collect backtraces from Info Table Provenance Entries
        IPEBacktraceMech
      | -- | use 'HasCallStack'
      HasCallStackBacktraceMech
      deriving (Eq, Show)

    enabledBacktraceMechanisms :: IORef [BacktraceMechanism]
    enabledBacktraceMechanisms = unsafePerformIO $ newIORef []
    {-# NOINLINE enabledBacktraceMechanisms #-}

    -- | Set how 'Control.Exception.throwIO', et al. collect backtraces.
    setEnabledBacktraceMechanisms :: [BacktraceMechanism] -> IO ()
    setEnabledBacktraceMechanisms = writeIORef currentBacktraceMechanisms

    -- | Returns the currently selected 'BacktraceMechanism's.
    getEnabledBacktraceMechanisms :: IO [BacktraceMechanism]

>     getEnabledBacktraceMechanisms = readIORef currentBacktraceMechanisms

Author: Indeed, I agree that this global mutable state is indeed a rather unfortunate wart. However, in my mind these functions should be used only very sparingly and I can't claim to be imaginative enough to foresee all possible applications here so I would be reluctant to preclude disabling of mechanisms.

A ``collectBacktrace`` primitive used by ``throwWithContext``
simply dispatches to the currently-selected ``BacktraceMechanism``\ s: ::

    module GHC.Exception.Backtrace where

    -- | Collect an 'ExceptionContext' containing backtraces from all enabled
    -- 'BacktraceMechanism's.
    collectBacktraces :: HasCallStack => IO ExceptionContext
    collectBacktraces = do
        mechs <- enabledBacktraceMechanisms
        mconcat `fmap` mapM collectBacktrace mechs

    -- | Collect a 'Backtrace' via the given 'BacktraceMechanism'.
    collectBacktraces' :: HasCallStack => BacktraceMechanism -> IO ExceptionContext

    module GHC.Exception where

    -- | Throw an exception. Exceptions may be thrown from purely
    -- functional code, but may only be caught within the 'IO' monad.
    -- Backtraces are collected using the backtrace mechanisms selected by
    -- 'GHC.Exception.Backtrace.enabledBacktraceMechanisms'.
    throwWithContext :: forall e a. (HasCallStack, Exception e)
                     => e -> ExceptionContext -> a
    throwWithContext e ctxt = do
        -- (implementation simplified for clarity)
        backtraces <- collectBacktraces

>         raise# (addContext (ctxt <> backtraces) (toException e))

Author: It certainly could. In hindsight, it's actually a bit perplexing that I didn't choose that form. Not sure what I was thinking when I wrote this.

>         raise# (addContext (ctxt <> backtraces) (toException e))

Author: On further reflection, I'm actually unsure of whether this combinator even pulls its weight given the complexity that it adds to the exception interfaces. Specifically, we currently have four throw variants:

> `throw` ⨉ {pure, IO} ⨉ {with backtrace, without backtrace}

Adding `WithContext` variants adds yet another dimension to this space. However, the same end can be accomplished (admittedly, slightly less efficiently) with `throw` and `withExceptionContext`.

Note that in order to provide ``HasCallStack`` backtraces we propose that a
``HasCallStack`` constraint be added to ``throw``, ``throwIO``, and similar
functions. Our prototype implementation suggests that this likely does not
carry a significant performance impact.

Since some users may want to explicitly opt out of backtrace collection when
throwing certain exceptions, we also propose to add non-backtrace-collecting
``throw`` variants: ::

    throwNoBacktrace   :: forall e a. (Exception e) => e -> a
    throwIONoBacktrace :: forall e a. (Exception e) => e -> a

    throwWithContextNoBacktrace   :: forall e a. (Exception e) => e -> ExceptionContext -> a
    throwIOWithContextNoBacktrace :: forall e a. (Exception e) => e -> ExceptionContext -> a

Examples
--------

User programs would typically call ``setEnabledBacktraceMechanisms`` during
start-up to select a backtrace mechanism appropriate to their usage: ::

    main :: IO ()
    main = do

>         setEnabledBacktraceMechanisms [HasCallStackBacktrace, ExecutionStackBacktrace]

Author: It means that a `HasCallStack` annotation will not be added to thrown exceptions. In practice I can think of few reasons why one would want to do this since you have already opted in to paying the cost of collecting the `HasCallStack` backtrace when writing the program.

>         setEnabledBacktraceMechanisms [HasCallStackBacktrace, ExecutionStackBacktrace]

Reviewer 6: 🤔 yeah, I would imagine wanting that to be enabled by default

global backtrace mechanism selection facility proposed here recognizes this but
it suffers from the usual drawbacks associated with global state: it does not
compose well and may result in surprising behavior when manipulated by more
than one actor.

Migration
---------

Unlike previous versions of this proposal, the change described above has
nearly no impact on existing user-code while allowing existing users to benefit
from backtraces. The only direct breakage will result in applications of the
``SomeException`` data constructor, where the user will be faced with a
compile-time error complaining that ``?exceptionContext`` is not in scope.
In our experience, this sort of code is rare and generally quite
straightforward to adapt; a survey of Hackage suggests that nearly all uses of
``SomeException`` are in pattern contexts.

One existing use-case which does not break but arguably results in non-ideal
behavior is that of exception re-throwing. For instance, consider the program:
::

    catch do_something $ \(e :: MyException) ->
        -- Do something
        throwIO e

Here the original annotations attached to ``e``  (which may include, e.g.,

> backtraces) will be lost when the exception is re-thrown.

Author: I have adopted this.

> backtraces) will be lost when the exception is re-thrown.

Reviewer 7: Looks like the proposal text still needs to be adapted now that this has been fixed?

Decorate exceptions with backtrace information
==============================================

.. authors:: Ben Gamari; David Eichmann; Sven Tennie
.. date-accepted::
.. ticket-url:: https://gitlab.haskell.org/ghc/ghc/issues/18159
.. implemented:: in-progress <https://gitlab.haskell.org/ghc/ghc/-/merge_requests/8869>
.. highlight:: haskell
.. header:: This proposal is `discussed at this pull request <https://github.com/ghc-proposals/ghc-proposals/pull/330>`_.

> .. contents::

Reviewer 7: This proposal clearly predates the addition of section numbers to the proposal template. :grin: 

```suggestion
.. sectnum::
.. contents::
```

by attaching backtraces to exceptions.

Motivation
----------
Exceptions are one of the primary mechanisms by which Haskell programs report
errors. However, in contrast to most languages, Haskell provides only few tools for
identifying the source of such errors. This leads to a poor debugging experience
and makes it difficult to monitor systems in production.

While over the past few years GHC has grown a variety of mechanisms for reporting
backtraces (e.g. ``HasCallStack``, ``GHC.Stack.CCS``, and DWARF debug
information), currently we do not have a means to attach such backtraces to
exceptions. The goal of this proposal is to fix this `long-standing
<https://www.youtube.com/watch?v=J0c4L-AURDQ>`_ problem.

We want to ensure that exceptions report provenance information *by
default* without requiring action on the part of the developer. To provide this provenance we leverage
the existing mechanisms for collecting backtraces listed above. Furthermore, we
want to ensure that this information is available for consumption in structured
form by the user program, to allow use by logging libraries (for instance, see

> `kaptip-raven #1

Reviewer 7: ```suggestion
`katip-raven #1
```

<https://github.com/cachix/katip-raven/issues/1#issuecomment-625389463>`_),
automatic error reporting, code analysis tools, and the like.

Proposed Change Specification
-----------------------------

> This proposal consists of three largely-independent parts:
> 
> 1. augmenting the current root of the exception hierarchy,
>    ``SomeException``, with additional metadata in the form of an
>    ``ExceptionContext``.
> 2. the introduction of annotation types to capture backtraces from the
>    above-described backtrace-collection mechanisms
> 3. a facility for choosing which backtrace mechanism(s)
>    should be used to collect exception provenance.

Reviewer 7: ```suggestion
This proposal consists of four largely-independent parts:

1. augmenting the current root of the exception hierarchy,
   ``SomeException``, with additional metadata in the form of an
   ``ExceptionContext``;
2. the introduction of annotation types to capture backtraces from the
   above-described backtrace-collection mechanisms;
3. a facility for choosing which backtrace mechanism(s)
   should be used to collect exception provenance;
4. changing the top level exception handler to use `displayException`
   rather than `show`.
```

``SomeException`` to carry an additional "context" value, consisting of a collection
of dynamically-typed annotations. The dynamic nature of these annotations
allows the mechanism to handle not just backtraces but any arbitrary
context-relevant value which may be inspected by a handler for analysis and logging.
To avoid breaking existing users of ``SomeException``, we introduce this
context as an implicit parameter constraint:  ::

    data SomeException where
        SomeException :: forall a. (Exception a, ?exceptionContext :: ExceptionContext)
                      => a -> SomeException

    data ExceptionContext = ExceptionContext [SomeExceptionAnnotation]

    -- | Contexts are merged by concatenation
    instance Semigroup ExceptionContext
    instance Monoid ExceptionContext

Following the example of the ``Exception`` class, we propose that exception
annotations implement a minimal typeclass which provides ``Typeable`` evidence
and the ability to render the annotation to a ``String``: ::

    data SomeExceptionAnnotation where
        SomeExceptionAnnotation :: forall a. (ExceptionAnnotation a)
                                => a -> SomeExceptionAnnotation

>     class Typeable a => ExceptionAnnotation a where

Reviewer 7: `Exception` has a `Show` superclass, should we do the same for `ExceptionAnnotation` for consistency? I guess using `DefaultSignatures` for `displayExceptionAnnotation` is fine too though.

Author: I suppose we could add a `Show` superclass although I am of mixed feelings as to whether it is desireable. Specifically, I see two viewpoints:

 1. That `Exception` requiring `Show` was a mistake in light of the fact that we have `displayException`. In fact, I believe that we only have it today because the original design of `Exception` didn't include a `displayException` method.
 2. That `Show` complements `displayException`, serving as a machine-readable rendering while `displayException` is human-readable.

Given that not all types have law-abiding `Show` instances, my tendency is to agree more with (1). Making `Show` a superclass of `ExceptionAnnotation` will only serve to force more non-law-abiding `Show` instances while carrying little benefit.

To allow users to populate this new annotation field we propose that the
``Exception`` class be extended with a new ``toExceptionWithContext`` method: ::

    class (Typeable e, Show e) => Exception e where
        -- These are unchanged:
        toException            :: e -> SomeException
        fromException          :: SomeException -> Maybe e
        displayException       :: e -> String

        -- This is new:
        toExceptionWithContext :: e -> ExceptionContext -> SomeException
        toExceptionWithContext e ?exceptionContext = SomeException e

        -- toException is implemented in terms of toExceptionWithContext
        toException e = toExceptionWithContext e mempty

The ``Exception`` instance for ``SomeException`` would accumulate and display
contexts: ::

    instance Exception SomeException where
        -- This is unchanged:
        toException se = se
        fromException = Just

        -- toExceptionWithContext *adds* context to an existing SomeException:
        toExceptionWithContext se ctxt = addExceptionContext ctxt se

        -- displayException shows context after the exception itself:
        displayException (SomeException e) =

>             displayException e ++ "\n" ++ displayExceptionContext ?exceptionContext

Reviewer 7: Even if the exception context is empty, this will add an extra trailing newline. Perhaps make `displayExceptionContext` add this only if the context is non-empty?

Author: Quite right. I shall fix this.

    displayExceptionContext :: ExceptionContext -> String
    displayExceptionContext = ...

Since the ``SomeException``'s ``displayException`` implementation is used to
by GHC's top-level exception handler to display uncaught exceptions, this
change carries the consequence that uncaught exceptions will have their context

> automatically printed as part of the error message presented to the user.

Reviewer 7: Perhaps add a forward reference to the section "Teach top-level handler to use displayException"?

mechanisms to be captured attached to exceptions.

Specifically, we propose to introduce the following set of exception
annotations: ::

    -- | A backtrace from the cost-centre profiler.
    data CostCentreBacktrace = CostCentreBacktrace { ... }
    instance ExceptionAnnotation CostCentreBacktrace

    -- | A backtrace using the native (e.g. DWARF-based) stack unwinder
    data ExecutionStackBacktrace = ExecutionStackBacktrace { ... }
    instance ExceptionAnnotation ExecutionStackBacktrace

    -- | A backtrace from GHC's Haskell stack unwinder and info-table
    -- provenance map.
    data InfoProvBacktrace = InfoProvBacktrace { ... }
    instance ExceptionAnnotation InfoProvBacktrace

    -- | A backtrace from HasCallStack evidence.
    data HasCallStackBacktrace = HasCallStackBacktrace { ... }
    instance ExceptionAnnotation HasCallStackBacktrace

Handling of rethrowing
^^^^^^^^^^^^^^^^^^^^^^

One pattern frequently seen in Haskell programs is *rethrowing*. Typically this
takes the form of catching one type of exception and throwing in its place
another exception more specific to the application domain. For instance,

> :: haskell

Reviewer 8: GitHub struggles to render this code snippet. Please adjust.

    readFile "my-app.conf" `catch` $ \ (ioe :: IOError) ->
        if isDoesNotExistError ioe
          then throwIO MissingConfigurationError
          else throwIO ioe

This pattern can be problematic in the presence of exception context: the
exception thrown by the handler lacks any of the context attached to the
original ``IOError``, including any backtraces.

While in some select cases dropping context may be desireable (e.g. to avoid
exposing implementation details unnecessarily to the user), in general this
proposal seeks to make exception provenance information ubiquitous and
reliable. Consequently, we propose to that ``catch`` and ``handle`` be modified

> to preserve exception context when an exception is thrown from a handler.

Reviewer 7: Can we see what these modifications look like?

In particular I'm wondering if `throwIO` inside `catch` could add an `ExceptionAnnotation` represented as a new "caused by" datatype that stores a `SomeException` (and hence its backtraces/annotations).

Selecting backtrace mechanisms
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

With the machinery described above, we can now address a common debugging
scenario: locating the origin of an exception thrown by a third-party library.
By far the most common means of throwing exceptions are ``throw``, ``throwIO``,
``error``, and ``undefined``, none of which have any notion of backtrace collection.
This raises the question of how the user should select which of the above
mechanism(s) these functions should use to collect their backtrace.

For this we propose a pragmatic, stateful approach to allow the user to enable
individual mechanism(s) should be used for backtrace collection in ``throw``,
``throwIO`` and similar functions: ::

    module GHC.Exception.Backtrace

>         ( enableMechanism, BacktraceMechanism(..) ) where

Reviewer 7: Is this export list complete? It doesn't mention `disableMechanism` or `collectBacktrace[s]`.

    -- | Which kind of backtrace to collect when an exception is thrown.
    data BacktraceMechanism
      = -- | collect a cost center stacktrace (only available when built with profiling)

>         CostCenterBacktraceMech

Reviewer 7: I would spell out `BacktraceMechanism` rather than slightly abbreviating it in the constructor names, personally.

      | -- | use execution stack unwinding with given limit
        ExecutionStackBacktraceMech
      | -- | collect backtraces from Info Table Provenance Entries
        IPEBacktraceMech
      | -- | use 'HasCallStack'
        HasCallStackBacktraceMech
      deriving (Eq, Show)

    -- | Enable the given 'BacktraceMechanism' to be used when
    -- 'Control.Exception.throwIO', et al. collect backtraces.
    enableMechanism :: BacktraceMechanism -> IO ()
    enableMechanism = ...
        -- Internally this would be mutate program-global state

    disableMechanism :: BacktraceMechanism -> IO ()
    disableMechanism = ...

A ``collectBacktrace`` primitive used by ``throwWithContext``
simply dispatches to the currently-selected ``BacktraceMechanism``\ s: ::

    module GHC.Exception.Backtrace where

    -- | Collect an 'ExceptionContext' containing backtraces from all enabled
    -- 'BacktraceMechanism's.
    collectBacktraces :: HasCallStack => IO ExceptionContext
    collectBacktraces = do

>         mechs <- readIORef currentBacktraceMechanisms

Reviewer 7: Perhaps expose an API function from `GHC.Exception.Backtrace` to query the current list of backtrace mechanisms?

carry a significant performance impact.

Since some users may want to explicitly opt out of backtrace collection when
throwing certain exceptions (e.g. in codebases where exceptions are used for
non-exceptional flow control), we also propose to add non-backtrace-collecting
``throw`` variants: ::

    throwNoBacktrace   :: forall e a. (Exception e) => e -> a
    throwIONoBacktrace :: forall e a. (Exception e) => e -> a

Teach top-level handler to use ``displayException``
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

For historical reasons, GHC's top-level exception handler currently displays
unhandled exceptions using ``Show`` rather than the ``Exception`` class's
``displayException``.

Since only ``displayException`` will display exception
context, we propose that this behavior is changed: unhandled exceptions
should be displayed to the user using ``displayException``.
As the default implementation of ``displayException`` simply delegates to
``show``, we expect that the messages produced by most exceptions will be
unaffected by this change (except for the context added by ``SomeException``\'s
``displayException`` implementation).

Examples
--------

> User programs would typically call ``setEnabledBacktraceMechanisms`` during

Reviewer 7: Looks like the current version of the proposal doesn't mention `setEnabledBacktraceMechanisms` in the API of `GHC.Exception.Backtrace`?

even if we don't provide every possible variant. Moreover, these changes will
need to be mirrored in downstream packages (e.g. ``exceptions``).

Moreover, the general nature of exception context slightly muddies the waters
when it comes to exception hierarchy design. Library authors now have two ways
of conveying failure information to the caller: they may introduce a new
exception type (as they can do today) or they can augment an existing exception
type via the context field. Correctly choosing from between these options may
be, in some cases, non-obvious and could require an element of design taste.

The introduction of the global state for backtrace mechanism selection is quite
ad-hoc. We consider this approach to be a compromise which makes robust
backtraces available by default with minimal additional code. Exception
backtraces are primarily a debugging tool and are a cross-cutting concern. The
global backtrace mechanism selection facility proposed here recognizes this but
it suffers from the usual drawbacks associated with global state: it does not
compose well and may result in surprising behavior when manipulated by more

> than one actor.

Reviewer 7: I think this is a reasonable, pragmatic compromise. Does the proposal mention what the default state for the enabled backtraces is if the user doesn't do anything? I couldn't immediately see this, but I might have missed it.

addressed by this proposal but would lose out on many of the benefits of
offering structured backtraces to the user, in addition to significantly
complicating implementation.

Handling of rethrowing
~~~~~~~~~~~~~~~~~~~~~~

The preservation of ``ExceptionContext`` in ``catch``, et al. is a design
choice whose value (namely, assurance context is not lost on rethrowing) may
not be worth the slight overhead it imposes.

In addition, there is the question of whether rethrown exceptions should gain a
backtrace for the ``catch`` callsite. We currently err on "no" here to avoid
undue overhead, but it may be worth revisiting this in the future.

Ubiquity of ``HasCallStack``
~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Today, ``HasCallStack`` is the most commonly available and therefore widely
used backtrace mechanism. However, it can introduce overhead by way of small
amounts of allocation in otherwise non-allocating code. The proposal above adds
a ``HasCallStack`` constraints to ``throw``.

Implementation Plan
-------------------

There is an active branch with an implementation of this proposal:
<https://gitlab.haskell.org/ghc/ghc/-/merge_requests/8869>

Acknowledgments
---------------

>  * Sven Tennie (``@supersven``) has been the driving force through most of this proposal, having

Reviewer 8: This is rendered as a quote because of the space before `*`.

  It never calls ``collectBacktraces``, adding **only** the user-specified
  annotation.

Providing context to handlers
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Export the following new definitions from ``Control.Exception`` which provide a
convenient way to gain access to ``ExceptionContext`` in exception handlers: ::

  data ExceptionWithContext a =
    ExceptionWithContext ExceptionContext a

  instance Show a => Show (ExceptionWithContext a)

  instance Exception a => Exception (ExceptionWithContext a) where
      toException (ExceptionWithContext ctxt e) = SomeException e
        where ?exceptionContext = ctxt
      fromException se = do
          e <- fromException se
          return (ExceptionWithContext (exceptionContext se) e)
      displayException = displayException . toException

Preserving context on rethrowing
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

In ``Control.Exception``, modify existing definitions as follows:

* Modify ``catch`` to preserve ``ExceptionContext`` when a handler rethrows an
  exception: ::

    catch :: Exception e => IO a -> (e -> IO a) -> IO a
    catch (IO io) handler = IO $ catch# io handler'
     where
       handler' e =
         case fromException e of

>            Just e' -> unIO (annotationIO (someExceptionContext e) (handler e'))

Author: Indeed, good catch!

           Nothing -> raiseIO# e

  Modify ``catchJust`` and ``handleJust`` accordingly (mutatis mutandis).

Capturing Backtraces on Exceptions
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

In ``Control.Exception``, modify existing definitions as follows:

* Add the following method and default definition to the ``Exception``
  typeclass: ::

    backtraceDesired :: forall a. Bool
    backtraceDesired = True

* Modify ``throwIO`` as follows: ::

>     throwIO :: forall e a. Exception e => e -> IO a

Author: Indeed, goal was to separate concerns. I've pointed this out explicitly.

    throwIO :: forall e a. Exception e => e -> IO a
    throwIO e = do
        ?exceptionContext <-
          if backtraceDesired e
            then do
              bt <- collectBacktraces
              return (addExceptionAnnotation bt emptyExceptionContext)
            else return emptyExceptionContext
        let se :: SomeException
            se = SomeException e
        raiseIO# se

Export the following new definitions from ``Control.Exception``:

* The following ``newtype`` wrapper and instance which can be used by the user
  when throwing an exception to disable backtrace collection: ::

    newtype NoBacktrace e = NoBacktrace e

    instance Show e => Show (NoBacktrace e)

    instance Exception e => Exception (NoBacktrace e) where
      fromException = NoBacktrace . fromException
      toException (NoBacktrace e) = toException e
      backtraceDesired _ = False

``HasCallStack`` Backtraces for Thrown Exceptions
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

In ``Control.Exception`` add ``HasCallStack`` constraints to the exception
``throw`` functions to allow inclusion in backtrace context: ::

    throwIO :: forall e a. (HasCallStack, Exception e) => e -> IO a

>     throw   :: forall e a. (HasCallStack, Exception e) => e -> a

Reviewer 4: I assume that you intend `throw` to collect backtraces like `throwIO` does, but this is not specified.

How about `error`?

Author: A very good point since `error` is not implemented in terms of `throw`.

This also raises the question of what to do about the `ErrorCallWithLocation` constructor, which carries a backtrace. In the interest of keeping the scope of the proposal at bey, I propose to leave it as-is currently. This means that there will be two backtraces produced by `error`: one in the `ExceptionContext` and another in the `ErrorCallWithLocation` constructor.

Author: A very good point since `error` is not implemented in terms of `throw`.

This also raises the question of what to do about the `ErrorCallWithLocation` constructor, which carries a backtrace. In the interest of keeping the scope of the proposal at bey, I propose to leave it as-is currently. This means that there will be two backtraces produced by `error`: one in the `ExceptionContext` and another in the `ErrorCallWithLocation` constructor.

Author: This also raises the question of what should happen with `errorWithoutStackTrace`. I will propose that we keep it unchanged for now, meaning that it will not carry a `Backtrace` in its `ExceptionContext`.

Reviewer 4: I think both are quite fair. At any rate, `error` is not typically called in a performance sensitive situation as it's more of a crash and burn type of thing.

Asynchronous exceptions
~~~~~~~~~~~~~~~~~~~~~~~

Modify the following definitions in ``GHC.Conc.Sync``: ::

>     throwTo :: forall e. (Exception e, HasCallSTack) => ThreadId -> e -> IO ()

Reviewer 9: typo: HasCallStack

     backtrace collection; imprecise backtraces; large binary size overhead
     (built with ``-g3``)
* Info-table provenance (IPE) information (via ``GHC.Stack.CloneStack``):
   * Pros: Can be used on all platforms; no runtime overhead
   * Cons: Large binary size overhead; no visibility into foreign code; must be
     built with ``-finfo-table-map``

All of these backtrace mechanisms have their uses, offering a range of levels
of detail, executable size, and runtime overhead. Given the complementary
nature of these mechanisms, GHC should not dictate which of these mechanisms
should be used to report exception backtraces. Consequently, we use the
above-described context mechanism to allow backtraces from any of these
mechanisms to be captured attached to exceptions.

The fact that backtrace collection with some of these mechanisms can be
rather expensive motivates two features of this proposal:

* the ``NoBacktrace`` wrapper, allowing users to disable backtrace collection
  at the ``throw``-site. This is sometimes necessary when exceptions are used
  for non-exceptional control flow.

* the ability to enable and disable individual exception mechanisms via

>   ``setEnabledBacktraceMechanisms``.

Reviewer 4: This name is used here and below, but in the specification, above, you use `setBacktraceMechanismState` instead. I think they're not of the same type either.

      SomeExceptionWithLocation
        :: forall e. Exception e
        => Maybe Backtrace   -- ^ backtrace, if available
        -> e                 -- ^ the exception
        -> SomeException

    pattern SomeException e <- SomeExceptionWithLocation _ e
      where
        SomeException e = mkSomeExceptionWithLocation e

The problem with this is that the pattern match completeness checker does not
play well with pattern synonyms. Additionally, it may introduce a ``MonadFail``
constraint where one previously did not exist. For example, the following would no
longer typecheck due to the lack of a ``MonadFail m`` constraint: ::

    f :: Monad m => SomeException -> m ()
    f someException = do
      SomeException e <- pure someException   -- Pattern synonym is assumed fallible
      ...

Backtrace mechanism selection
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In addition, there are several alternatives to the proposed backtrace mechanism
selection facility. For instance:

* a simpler, non-GADT-based approach might be used
* GHC could gain support for setting the backtrace mechanism at compile-time
  via a compiler flag (this would essentially come down to GHC emitting a call

>   to ``enabledBacktraceMechanisms`` in its start-up code).

Reviewer 10: Why not have GHC default to enabling all backtrace mechanisms that might contain information? I think with the exception of `HasCallStack` they're all enabled by compiler flags, so it should be easy for GHC to know which mechanisms to enable.

I can see a perf concern, but between

1. disabling `backtraceDesired` for control-flow exceptions, and
2. the fact that the expensive mechanisms seem unlikely to be enabled outside of specific debugging builds

the concern seems surmountable to me.

Author: > so it should be easy for GHC to know which mechanisms to enable.

Unfortunately this is true only for the cost-centre and `HasCallStack` mechanisms. Both IPE and DWARF are trickier in that they are not "global": some modules may be built with IPE/DWARF information while others may not.

To make matters worse, IPE mechanism performs its metadata lookups lazily: it freezes the stack when the exception is thrown (essentially performing a potentially expensive copy) and only tries to lookup the associated IPE entries when someone tries to examine the backtrace. This means that users enabling the IPE mechanism will end up incurring the cost of a stack copy regardless of whether they have IPE information available. Technically we could use heuristics (e.g. checking whether the IPE table has no entries) to disable IPE backtrace collection but I'm reluctant to do this since they would be hard to predict and easily defeated (e.g. if your compiler was built with IPE information but your application was not).

  data ExceptionWithContext a =
    ExceptionWithContext ExceptionContext a

  instance Show a => Show (ExceptionWithContext a)

  instance Exception a => Exception (ExceptionWithContext a) where
      toException (ExceptionWithContext ctxt e) = SomeException e
        where ?exceptionContext = ctxt
      fromException se = do
          e <- fromException se
          return (ExceptionWithContext (someExceptionContext se) e)
      displayException = displayException . toException

Preserving exception causes on rethrowing
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

In ``Control.Exception``:

* Introduce a ``newtype``: ::

    newtype WhileHandling = WhileHandling SomeException

    instance ExceptionAnnotation WhileHandling

* Modify ``catch`` to add ``WhileHandling`` annotations to exceptions thrown from handlers: ::

    catch :: Exception e => IO a -> (e -> IO a) -> IO a
    catch (IO io) handler = IO $ catch# io handler'
     where
       handler' e =
         case fromException e of
           Just e' -> unIO (annotateIO (WhileHandling e) (handler e'))
           Nothing -> raiseIO# e

  Modify ``catchJust`` and ``handleJust`` accordingly (mutatis mutandis).

> * Introduce ``catchNoAnnotation`` to exposing the old semantics of ``catch``: ::
> 
>     catchNoAnnotation :: Exception e => IO a -> (e -> IO a) -> IO a
>     catchNoAnnotation (IO io) handler = IO $ catch# io handler'

Reviewer 11: ```suggestion
* Introduce ``catchNoAnnotation`` exposing the old semantics of ``catch``: ::

    catchNoAnnotation :: Exception e => IO a -> (e -> IO a) -> IO a
    catchNoAnnotation (IO io) handler = IO $ catch# io handler'
```

     where
       handler' e =
         case fromException e of
           Just e' -> unIO (handler e')
           Nothing -> raiseIO# e

In ``GHC.IO``:

> * Introduce ``catchNoAnnotation`` to exposing the old semantics of ``catch``: ::
> 
>     catchExceptionNoAnnotation :: Exception e => IO a -> (e -> IO a) -> IO a
>     catchExceptionNoAnnotation !io handler = catchNoAnnotation io handler

Reviewer 11: ```suggestion
* Introduce ``catchExceptionNoAnnotation`` exposing the old semantics of ``catch``: ::

    catchExceptionNoAnnotation :: Exception e => IO a -> (e -> IO a) -> IO a
    catchExceptionNoAnnotation !io handler = catchNoAnnotation io handler
```

    throwIO e = do
        se <- toExceptionWithBacktrace e
        raiseIO# se

* Modify ``throw`` similarly: ::

    throw :: forall (r :: RuntimeRep). forall (a :: TYPE r). forall e.
             (?callStack :: CallStack, Exception e) => e -> a
    throw e =
        let !se = unsafePerformIO (toExceptionWithBacktrace e)
        in raise# se

* Modify ``GHC.Exception.errorCallWithCallStackException`` to use
  ``toExceptionWithBacktrace`` instead of ``toException``. This ensures that
  ``error`` and ``undefined`` gain ``Backtraces``.

Export the following new definitions from ``Control.Exception``:

* The following ``newtype`` wrapper and instance which can be used by the user
  when throwing an exception to disable backtrace collection: ::

    newtype NoBacktrace e = NoBacktrace e

    instance Show e => Show (NoBacktrace e)

    instance Exception e => Exception (NoBacktrace e) where
      fromException = NoBacktrace . fromException
      toException (NoBacktrace e) = toException e
      backtraceDesired _ = False

In ``GHC.IO``:

* Modify ``onException`` to avoid capturing a new backtrace: ::

    onException :: IO a -> IO b -> IO a

>     onException io what = io `catchExceptionNoAnnotate` \e -> do

Reviewer 7: ```suggestion
    onException io what = io `catchExceptionNoAnnotation` \e -> do
```

context as an implicit parameter constraint:  ::

    data SomeException where

>         SomeException :: forall a. (Exception a, ?context :: ExceptionContext)

Reviewer 8: This could conflict with an application-specific implicit parameter named `?context`. Why not keep it as `?exceptionContext`?

Author: Sigh, I could have sworn that there was a suggestion to change this to `?context` earlier in the discussion and which I somewhat mindlessly followed. However, I share your concern; I'll change this to `exceptionContext`.
